### PR TITLE
RichText: Add a format prop to allow HTML string values to be used in RichText components

### PR DESCRIPTION
**Pull request by youknowriad**
We discussed several times whether we want to treat RichText values as strings, elements or raw trees. In this PR I'm introducing a `format` prop for the RichText component for two reasons:

 - If someone wants to store HTML in a meta attribute, the RichText component should produce a string output and not an array of element. This has been raised a while ago but I'm not able to find the corresponding issue.

 - We know we want to move away from the element's tree as default representation of RichText value (because it's not serializable properly on the server), we're not certain yet about the alternative. But anyway in order to do so (change the format) we need a migration plan, this `format` will allow us to introduce warnings and at some point switch the default `format` (which is `element` at the moment) and at a later point maybe remove the `element` format entirely.

**Testing instructions**

 - Right now, the newly introduce "string" format is not used anywhere, so this should not change anything for the existing blocks

 - You can setup a meta block with a "string" format for the RichText component.

### CODE REVIEW

File: blocks/rich-text/index.js
Comment (iseulde) [inline blocks/rich-text/index.js]: Will this be executed many times? This will be much slower than `this.editor.getContent( { format: 'raw' } )` or one loop over `this.editor.getBody().childNodes`.

**Comment by iseulde** [comment]
Just wondering, but what's the point of adding it if it is used anywhere?

File: blocks/rich-text/index.js
Comment (youknowriad) [reply blocks/rich-text/index.js]: It will be executed on each change (type), It's probably slower but since it's limited to a single RichText, maybe it's not that important.

`raw` is not great because we don't want to save the TinyMCE temporary nodes and attributes. We could probably write an alternative using `childNodes` and stripping these nodes, but not sure if this is different of what's done internally in TinyMCE

**Comment by youknowriad** [comment]
@iseulde As a said in the description, it's an extensibility API that have been asked for to save HTML to meta.

And it could be used on a follow-up PR as alternative to the `children` source for some blocks.

**Comment by iseulde** [comment]
Would it be okay to add a use case here to demonstrate and to ensure it works as expected? Cool if it's too complicated, but that might be a sign more work is needed elsewhere if it's not just a one line change.

**Comment by youknowriad** [comment]
Sure, here's a small patch to apply to test it on the Audio block (caption)


```diff
diff --git a/blocks/library/audio/index.js b/blocks/library/audio/index.js
index 803bc907f..0565d0b3e 100644
--- a/blocks/library/audio/index.js
+++ b/blocks/library/audio/index.js
@@ -13,7 +13,7 @@ import {
        Placeholder,
        Toolbar,
 } from '@wordpress/components';
-import { Component } from '@wordpress/element';
+import { Component, RawHTML } from '@wordpress/element';
 import { mediaUpload } from '@wordpress/utils';
 
 /**
@@ -44,8 +44,8 @@ export const settings = {
                        attribute: 'src',
                },
                caption: {
-                       type: 'array',
-                       source: 'children',
+                       type: 'string',
+                       source: 'html',
                        selector: 'figcaption',
                },
                id: {
@@ -161,6 +161,7 @@ export const settings = {
                                                        onChange={ ( value ) => setAttributes( { caption: value } ) }
                                                        isSelected={ isSelected }
                                                        inlineToolbar
+                                                       format="string"
                                                />
                                        ) }
                                </figure>,
@@ -174,7 +175,7 @@ export const settings = {
                return (
                        <figure>
                                <audio controls="controls" src={ src } />
-                               { caption && caption.length > 0 && <figcaption>{ caption }</figcaption> }
+                               { caption && caption.length > 0 && <figcaption><RawHTML>{ caption }</RawHTML></figcaption> }
                        </figure>
                );
        },
```

File: blocks/rich-text/format.js
Comment (aduth) [inline blocks/rich-text/format.js]: Minor: I don't think we try to align `return` and `param` descriptions now that they're separated with newline.

File: blocks/rich-text/format.js
Comment (aduth) [inline blocks/rich-text/format.js]: This function gives me bad vibes, for two reasons:

- The implementation is susceptible to XSS
   - As used, probably not in a way that would be of any practical concern
- We're performing two levels of transformation, first from string to DOM, then from DOM to element

Since its use is quite limited (for default value), maybe not of huge concern at the moment.

Comment (aduth) [reply blocks/rich-text/format.js]: Though it makes me wonder, why don't we just apply the default value as `dangerouslySetInnerHTML` ?

File: blocks/rich-text/index.js
Comment (aduth) [inline blocks/rich-text/index.js]: There's a lot of ternaries in this file now and it seems like it'd be easy to make mistakes in future maintenance.

I wonder if a simple improvement might just be to encapsulate logic to a single function:

```js
domToFormat( nodes, format )
```

**Review by aduth — COMMENTED** [review]
Through my various failures in trying to drop the `children` source (#5380), I ultimately landed at the same conclusion that this type of format option would be the best path for those who need to work with content as an HTML string, so I'm generally on board with it. I think it does raise the maintenance burden in needing to support multiple formats, particularly when none of the core blocks have shown to have need for it.

One thing which I considered in #5380 is the idea of a `RichText.Content` component for use in save, so that a developer doesn't need to do the `RawHTML` rendering themselves. Do you have any thoughts on this? I guess if they need to explicitly assign `format`, there's already some awareness of the data format. And it would be unfortunate if this needed to be repeated for the `RichText.Content` element (unless we otherwise "infer" format from the shape of the value).

File: blocks/rich-text/format.js
Comment (youknowriad) [reply blocks/rich-text/format.js]: Good idea it allows us to factorize more logic using `valueToString(value, format)`

**Comment by youknowriad** [comment]
> One thing which I considered in #5380 is the idea of a RichText.Content component for use in save, so that a developer doesn't need to do the RawHTML rendering themselves. Do you have any thoughts on this? I guess if they need to explicitly assign format

I actually like the `RichText.Content` and I think we should consider using it even if the value is an element (for consistency). But I personally prefer explicitly assigning the format if it's different that the default format to avoid ambiguities. For example an element could be an object and a potential tree format is an object as well.

Let's introduce `RichText.Content` in a separate PR

**Comment by youknowriad** [comment]
Can we get this in?

File: blocks/rich-text/index.js
Comment (iseulde) [inline blocks/rich-text/index.js]: Any way to prevent the empty check from running twice here?

File: blocks/rich-text/index.js
Comment (iseulde) [reply blocks/rich-text/index.js]: Yeah, I think this is fine as long as it's used on inline `RichText` instances. Once we get lists to move over to nested blocks we can even get rid of multiline instances altogether.

File: blocks/rich-text/format.js
Comment (iseulde) [inline blocks/rich-text/format.js]: In `domToElement`, we're always stripping any TinyMCE internal element and attributes, but here we're not, even though there might be cases where we receive some in the case of splitting.

File: blocks/rich-text/test/__snapshots__/format.js.snap
Comment (iseulde) [inline blocks/rich-text/test/__snapshots__/format.js.snap]: Hm, isn't this invalid HTML? Won't React warn on nesting paragraphs? Or do we not care here for these tests?

File: blocks/rich-text/index.js
Comment (youknowriad) [reply blocks/rich-text/index.js]: Yes, maybe it could be inferred from the content.

File: blocks/rich-text/index.js
Comment (aduth) [inline blocks/rich-text/index.js]: Aside: Is the `isEmpty` check here meant to be an optimization for the easy case of generating an empty string? Is `isEmpty` itself actually performant? Wondering if we're actually shooting ourselves in the foot with this for the more common case that `RichText` is non-empty.

File: blocks/rich-text/index.js
Comment (aduth) [inline blocks/rich-text/index.js]: Aside: In the future, I might imagine we avoid working with the DOM as a source and instead defer to TinyMCE's `getContent( { format: 'tree' } )` to convert from their tree form to the desired target format. This would avoid the need for us to maintain our own cleaning logic. At which point, `domToFormat` doesn't make as much sense, might just be `toFormat( value, targetFormat, sourceFormat )` (maybe infer source format?).

File: blocks/rich-text/test/__snapshots__/format.js.snap
Comment (aduth) [reply blocks/rich-text/test/__snapshots__/format.js.snap]: Our new non-React serializer is much less noisy about issues it encounters. Not to say we shouldn't be using valid HTML here though.

**Review by aduth — APPROVED** [review]
This feels like a good refactoring of RichText, even without the new feature. Nice one 👍

File: blocks/rich-text/index.js
Comment (youknowriad) [reply blocks/rich-text/index.js]: It's not an optimization, it's to ensure the paragraph block is considered empty consistently. If we remove it we can end up with content like `['']` which will break the default appender.

File: blocks/rich-text/index.js
Comment (youknowriad) [reply blocks/rich-text/index.js]: I tend to always prefer explicitness over implicitness. I don't really care about the name of the function but I tend to be suspicious (maybe I shoudn't) every time I hear "infer from value"

File: blocks/rich-text/index.js
Comment (aduth) [reply blocks/rich-text/index.js]: I get the sense this has been discussed before, but if we have control over the elements we're producing, why are we keeping the empty string ?

File: blocks/rich-text/index.js
Comment (aduth) [reply blocks/rich-text/index.js]: Eh, I can see it, but it's kinda like if one had to pass a second argument to e.g. `Array.from`:

```js
Array.from( new Set( [ 1, 2, 3 ] ), 'Set' );
```

... one could say: "Well, duh, you passed me a set, I know it's a set"

File: blocks/rich-text/format.js
Comment (youknowriad) [reply blocks/rich-text/format.js]: Not sure how to address this in an optimal way. Don't we have tools in TinyMCE to do this for us instead of adding custom logic? Is it a requirement since we're not using this when getting the whole content of the editor but only when splitting?

File: blocks/rich-text/index.js
Comment (youknowriad) [reply blocks/rich-text/index.js]: I can't tell honestly :) maybe @iseulde can tell more here.

File: blocks/rich-text/index.js
Comment (youknowriad) [reply blocks/rich-text/index.js]: But sometimes it's not obvious, the easy example is that the tree structure and the element structure can't be distinguished easily because both can be plain objects.

File: blocks/rich-text/format.js
Comment (aduth) [reply blocks/rich-text/format.js]: `( new tinymce.dom.Serializer( {} ) ).serialize( el )` ? (Likely with some caching of a serializer)

Comment (youknowriad) [reply blocks/rich-text/format.js]: Thanks Andrew but unfortunately it's not working in my testing, we more likely have to specify the same rules used by TinyMCE internally

File: blocks/rich-text/index.js
Comment (iseulde) [reply blocks/rich-text/index.js]: > but if we have control over the elements we're producing

Do we have control over that? This one is getting the nodes from a contenteditable field which can contain all sorts combination. I've seen `<p></p>`,`<p>''</p>`, `<p><br></p>`, `<p><br>''</p>`, `<p><strong></strong></p>` etc., all of which are empty.

File: blocks/rich-text/format.js
Comment (iseulde) [reply blocks/rich-text/format.js]: > Is it a requirement since we're not using this when getting the whole content of the editor but only when splitting?

Well, the content will end up getting passed to the store, so I guess so. On consecutive edits we may pass `getContent` again, but this doesn't necessarily happen. Say you split a paragraph in two and the first part contains some elements with internal attributes. This will be passed on eventually to (block) `setAttributes`.

Not sure immediately how to fix either but I guess @aduth goes in the right direction. Maybe parse and serialise if you want to work on the string? I'll have a look if we can work on the fragment.

Comment (iseulde) [reply blocks/rich-text/format.js]: Did a little test here: https://github.com/WordPress/gutenberg/commit/39d9966a624eb625cd5ffc7cd6db433424bf73da

Splitting is currently entirely broken with the HTML format. The diff fixes it and should also get rid of internal stuff.

Comment (iseulde) [reply blocks/rich-text/format.js]: After the splitting fix, but without serialising, you'd get this:

```
The goal of this new editor is to make adding rich content to WordPress simple and enjoyable. This whole post is composed of <em data-mce-selected="inline-boundary">pieces of</em>
```

when splitting the first paragraph of the demo content.

**Review by ellatrix — CHANGES_REQUESTED** [review]
Looks good to me, but would love to see issues in https://github.com/WordPress/gutenberg/pull/6034#discussion_r181349661 fixed before merge.

**Comment by iseulde** [comment]
@youknowriad Fixed the splitting. There are still errors on merging though, where the merge function seems to assume React values.

**Comment by youknowriad** [comment]
@iseulde I tried splitting and merging and it works well for me with the last changes. The `merge` function is assume React values and that's fine because it's defined on the block level and the block author can update its behavior depending on the format used. I tried updating the function for paragraph block and it works great for me.

**Comment by iseulde** [comment]
Oh, I didn't update that. Sorry. Looks good to go then.

**Review by ellatrix — APPROVED** [review]

**Comment by zgordon** [comment]
Just want to throw in here that this would be super helpful and something have gotten requested from VIP.  Especially for migrating data from ACF or CMB2 fields and porting them over to blocks, or at least saving data in a more consistent way with what folks are used to.  Would love to see this rolled it, have been struggling with workarounds for this exact thing.  Thanks @youknowriad!

**Comment by aduth** [comment]
Noting that our `patterns` transforms make assumptions that the format would be of the `children` shape. Would this be unexpected for someone who otherwise uses the HTML format for a RichText element?

https://github.com/WordPress/gutenberg/blob/751a2fcdd7952fd4359984af292cc669cee065c8/packages/editor/src/components/rich-text/patterns.js#L193-L196

**Pull request merged**
